### PR TITLE
[ci] Fix SFT Real-PG numerics and restore full 8-GPU trigger

### PR DESCRIPTION
**Pull request by wwwjn**
Fixes the SFT numerics failure in https://github.com/pytorch/torchtitan/actions/runs/32751406256.

The checked-in Real-PG golden was generated before #4121 folded the physical batch dimension into the token sequence. For SFT, this changed First-Fit packing from eight independent 2,048-token bins to one 16,384-token bin. First-Fit preserves whole variable-length chat documents, so the larger bin changes packing efficiency, padding, and which supervised assistant tokens appear in each training step even though the nominal token budget remains 8 * 2,048. In a local two-rank reproduction, step 1 changed from 8,161 to 8,243 non-ignored labels. Position resets and document masks still prevent cross-document attention, but the training batch is not numerically identical. Reproducing immediately before and after #4121 changes step-1 loss from approximately 7.891477 to 7.889892.

The failed CI run completed all 10 training steps; only the exact golden comparison failed.

## Why multiple Real-PG goldens changed

The model goldens introduced by #4127 were generated by run 32097749288 on August 18, before #4121 landed on August 20. Although #4127 merged after #4121, its checked-in goldens still described the old physical `[local_batch_size, seq_len]` input layout.

#4121 preserves the per-rank token budget but folds the physical batch dimension into one flat sequence. For configurations whose old `local_batch_size` was greater than one, the data loader now packs documents into one larger token bin instead of several independent sequence-length bins. Document masks and position resets preserve attention semantics, but packing efficiency, padding, document selection, and therefore the exact tokens in a training step can change. Exact loss and gradient-norm goldens must consequently be regenerated. Two separate current-main Real-PG runs produced identical traces for Llama 3, DeepSeek V3, GPT-OSS, and Muse/Glimmer, confirming that those changes are stable stale-golden updates rather than run-to-run noise. The existing Llama 3 and GPT-OSS pipeline-parallel goldens remained bitwise unchanged and were not updated.

Qwen3 and Qwen3.5 exposed a separate determinism issue in their Real-PG EP numerics configurations. The default SelectiveAC policy forcibly recomputes the `moe.router.gate` matmul while caching `topk` results and expert-parallel `all_to_all` outputs. Backward replay therefore mixes recomputed router scores with routing decisions and communication results saved from the original forward. Repeated runs showed that this combination is not bitwise reproducible with real EP: step-1 forward loss can match while step-1 gradient norm differs, causing subsequent losses to diverge after the optimizer update.

The control matrix isolated the interaction:

- default SelectiveAC + real EP: not bitwise reproducible;
- SelectiveAC without forced router recomputation + real EP: bitwise reproducible;
- FullAC + real EP: bitwise reproducible;
- default SelectiveAC with EP=1: bitwise reproducible.

Each Qwen numerics test now uses one shared configuration for Fake-PG and Real-PG. That configuration retains SelectiveAC but sets `force_recompute_mm_shapes_by_fqns=[]`, keeping router scores, cached routing decisions, and cached EP communication outputs from the same forward execution. The default model configurations and dedicated SAC feature coverage are unchanged.

DeepSeek V3 `FSDP2+TP2+CP2+EP8` and Muse Glimmer `FSDP2+TP2+CP2` are now explicit Real-PG-only numerics cases. Direct Fake-PG probes of those configurations fail the finite-value guard at steps 1 and 2 respectively, so this PR removes the mode-specific `fake_pg_numerics_config` override mechanism instead of silently substituting different parallelism under the same test.

## This PR

- updates affected Real-PG loss and gradient-norm goldens;
- uses deterministic Qwen numerics recipes without forced router-gate recomputation in both Fake-PG and Real-PG;
- makes the DeepSeek and Muse CP numerics cases explicitly require Real-PG and removes `fake_pg_numerics_config`;
- restores the `ciflow/8gpu/*` tag trigger removed by #4127;
- documents that applying the `ciflow/8gpu` PR label runs every enabled feature and model test on the 8-GPU Real-PG lane;
- splits Real-PG into separate feature and model jobs, for both the PR-required subset and the full suite, while retaining the 60-minute per-job timeout;
- uses the existing explicit `--test_scope=real_pg_required` flag to select the smaller PR suite; tagged, main, and scheduled runs omit that flag and run the full suite;
- distinguishes required-subset and full-suite jobs in their check names.

### CODE REVIEW

File: .github/workflows/integration_test.yaml [hunk 1]
```diff
@@ -15,6 +15,8 @@ on:
         default: fake_pg
   push:
     branches: [ main ]
+    tags:
+      - ciflow/8gpu/*
```
Comment (tianyu-l) [inline .github/workflows/integration_test.yaml:19]: > documents that applying the ciflow/8gpu PR label runs every enabled feature and model test on the 8-GPU Real-PG lane;

ciflow/8gpu used to be reserved for rocm test. Is this changed to running both cuda and rocm on the real pg sets in features and models?

Comment (wwwjn) [reply .github/workflows/integration_test.yaml:19]: No, I reuse the `ciflow/8gpu` name and labeling strategy, and rocm test now is still disabled (I'd propose we have another tag `ciflow/rocm` for triggering rocm test, as they don't run in separate workflow with CUDA). 

Now the tag is used:
- If a PR is tagged with `ciflow/8gpu`, then every push on this PR will trigger a "8GPU real PG" run on **full** set of feature + model tests. 

The use case is for user who expected to change numerics and export the numerics before the PR is merged.

This is only used to trigger 8GPU real pg sets in features and models, separately in 2 runs (because putting them in one run will causing timeout of 120min limit)

Comment (tianyu-l) [reply .github/workflows/integration_test.yaml:19]: I think there was some reason why we used `ciflow/8gpu` instead of `ciflow/rocm` for rocm CI. Do you still remember? @akashveramd @huydhn

Comment (wwwjn) [reply .github/workflows/integration_test.yaml:19]: Thank you, I will wait for them to reply. My understanding is previous rocm and cuda runs are triggered in the same workflow and run at the same time if a PR is tagged by`ciflow/8gpu`, it doesn't means "trigger rocm" only. But I will wait and double check.

Comment (akashveramd) [reply .github/workflows/integration_test.yaml:19]: @tianyu-l @wwwjn: I believe this was where the discussion started for ciflow/8gpu label https://pytorch.slack.com/archives/C0A0E8ARXPZ/p1764843858514179. However, ciflow/rocm wasn't discussed. 
In my PR https://github.com/pytorch/torchtitan/pull/4039 we replaced ciflow/8gpu label with ciflow/rocm as that was preferred to run ROCm workflows.

**Review by tianyu-l — APPROVED** [review]

File: tests/assets/losses/real_pg/muse_glimmer_a10g.txt [hunk 2]
```diff
@@ -4,11 +4,11 @@
 # step loss grad_norm
 1 7.624619007110596 0.057667944580316544
 2 7.624410629272461 0.057388924062252045
-3 7.621224880218506 0.31297963857650757
```
Comment (wwwjn) [inline tests/assets/losses/real_pg/muse_glimmer_a10g.txt:7]: These changes in golden loss is because of removing the batch size dim PR. During my last rebase, it failed to be caught because 8 GPU real pg full suite is not triggered. Update in this PR

File: tests/integration_tests/models.py [hunk 3]
```diff
@@ -121,13 +128,16 @@ def build_model_tests_list() -> list[OverrideDefinitions]:
...
+            configs=[recipes.qwen35_debugmodel_moe_fsdp4_tp2_ep4_no_router_recompute],
             test_descr="Qwen3.5 MoE FSDP+TP+EP",
             test_name="qwen3_5_moe_fsdp+tp+ep",
             ngpu=8,
             golden_numerics_path=(
                 "tests/assets/losses/{execution_mode}/qwen3_5_a10g.txt"
             ),
+            # Forced router recomputation is not bitwise reproducible when
+            # mixed with cached Real-PG expert-parallel collectives.
+            fake_pg_numerics_config=recipes.qwen35_debugmodel_moe_fsdp4_tp2_ep4,
```
Comment (tianyu-l) [inline tests/integration_tests/models.py]: this field is making things not straightforward

Comment (wwwjn) [reply tests/integration_tests/models.py]: Oh sorry let me consolidate these 2 configs. For one test, they should use the same config, which is not recomputing the router.

Comment (wwwjn) [reply tests/integration_tests/models.py]: Consolidated in a277ced06. Both Qwen numerics tests now use one config for Fake-PG and Real-PG, and that shared config disables forced router recomputation. I removed the Qwen fake_pg_numerics_config overrides and the duplicate *_no_router_recompute recipe names.

Comment (tianyu-l) [reply tests/integration_tests/models.py]: but in general we still have https://github.com/pytorch/torchtitan/blob/main/tests/integration_tests/__init__.py#L41
Is it really necessary?

File: .github/workflows/integration_test.yaml (hunk 1 above)
Comment (jithunnair-amd) [reply .github/workflows/integration_test.yaml:19]: Actually, that's the direction we are planning to take as well: use the `ciflow/rocm` label to explicitly trigger ROCm CI on ROCm-specific workflows. So you can feel free to repurpose `ciflow/8gpu` as you see fit.

File: tests/integration_tests/models.py (hunk 3 above)
Comment (wwwjn) [reply tests/integration_tests/models.py]: No, not neccessary. Let me remove it. Now it's mistakenly used in DeepSeek test to run different parallelism for  numerics test on real_pg and fake_pg. Let me consolidate that

File: .github/workflows/integration_test.yaml (hunk 1 above)
Comment (jithunnair-amd) [reply .github/workflows/integration_test.yaml:19]: Although, @tianyu-l is also correct that the original use of `ciflow/8gpu` label was to [specifically turn on ROCm testing](https://github.com/pytorch/torchtitan/blob/4610fcab51c90385d07f2d41e87e51f1e2f8b486/.github/workflows/set-matrix.yaml#L112), which is also why I like the `ciflow/rocm` naming better, because it is more explicit.

File: tests/integration_tests/models.py (hunk 3 above)
Comment (tianyu-l) [reply tests/integration_tests/models.py]: sgtm, feel free to merge

Comment (wwwjn) [reply tests/integration_tests/models.py]: Sure, I will make sure everything works and merge soon

**Review by weifengpy — APPROVED** [review]
pray for landing (rocket gif)

**Comment by wwwjn** [comment]
> pray for landing (rocket gif)

Apologies for the delay, the whole github Action is done and I can not get the CI running result

**Comment by weifengpy** [comment]
> Apologies for the delay, the whole github Action is done and I can not get the CI running result

no worries - it's good that you are on this
